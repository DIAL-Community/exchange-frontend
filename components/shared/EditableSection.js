import { useIntl } from 'react-intl'
import { useState, useEffect, useCallback } from 'react'
import { FaSpinner } from 'react-icons/fa'
import EditButton from '../shared/EditButton'
import CreateButton from './CreateButton'

const EditableSection = ({
  canEdit,
  sectionHeader,
  editModeBody,
  displayModeBody,
  isDirty,
  isMutating,
  onSubmit,
  onCancel,
  createAction = null
}) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const [isInEditMode, setIsInEditMode] = useState(false)

  const [isSubmitInProgress, setIsSubmitInProgress] = useState(false)

  useEffect(() => {
    // if mutation has finished switch back to display mode
    if (isSubmitInProgress && !isMutating) {
      setIsInEditMode(false)
      setIsSubmitInProgress(false)
    }
  }, [isMutating]) // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <div className='mt-12' data-testid='editable-section'>
      <div className='flex justify-between items-center mb-3'>
        <div className='card-title text-dial-gray-dark'>{sectionHeader}</div>
        <div className='gap-3'>
          {canEdit && !isInEditMode &&
            <EditButton onClick={() => setIsInEditMode(true)} className='mr-2' />
          }
          {createAction && !isInEditMode &&
            <CreateButton label={format('app.create-new')} onClick={createAction} className='mr-2' />
          }
        </div>
      </div>
      {isInEditMode
        ? (
          <div className='bg-edit p-6'>
            {editModeBody}
            <div className='flex justify-end mt-8 gap-3 text-xl'>
              <button
                type='submit'
                onClick={() => {
                  setIsSubmitInProgress(true)
                  onSubmit()
                }}
                className='submit-button'
                disabled={!isDirty || isSubmitInProgress}
                data-testid='submit-button'
              >
                {format(`${isSubmitInProgress ? 'app.submitting' : 'app.submit'}`)}
                {isSubmitInProgress && <FaSpinner className='spinner ml-3 inline' data-testid='submit-spinner' />}
              </button>
              <button
                type='button'
                onClick={() => {
                  onCancel()
                  setIsInEditMode(false)
                }}
                className='cancel-button'
                disabled={isSubmitInProgress}
                data-testid='cancel-button'
              >
                {format('app.cancel')}
              </button>
            </div>
          </div>
        ) : displayModeBody
      }
    </div>
  )
}

export default EditableSection
