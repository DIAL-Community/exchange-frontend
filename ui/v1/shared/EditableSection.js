import { useIntl } from 'react-intl'
import { useState, useEffect, useCallback } from 'react'
import { FaSpinner } from 'react-icons/fa'
import EditButton from './form/EditButton'
import CreateButton from './form/CreateButton'

const EditableSection = ({
  canEdit,
  sectionHeader,
  editModeBody,
  displayModeBody,
  isDirty,
  isMutating,
  onSubmit,
  onCancel,
  createAction = null
}) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const [isInEditMode, setIsInEditMode] = useState(false)
  const [isSubmitInProgress, setIsSubmitInProgress] = useState(false)

  useEffect(() => {
    // if mutation has finished switch back to display mode
    if (isSubmitInProgress && !isMutating) {
      setIsInEditMode(false)
      setIsSubmitInProgress(false)
    }
  }, [isMutating]) // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <div className='flex flex-col gap-y-3'>
      <div className='flex justify-between items-center'>
        {sectionHeader}
        <div className='gap-3'>
          {canEdit && !isInEditMode &&
            <EditButton onClick={() => setIsInEditMode(true)} className='mr-2' />
          }
          {createAction && canEdit && !isInEditMode &&
            <CreateButton label={format('app.create-new')} onClick={createAction} className='mr-2' />
          }
        </div>
      </div>
      {isInEditMode
        ? (
          <div className='bg-edit'>
            {editModeBody}
            <div className='px-6 py-4 flex justify-end gap-3'>
              <button
                type='submit'
                onClick={() => {
                  setIsSubmitInProgress(true)
                  onSubmit()
                }}
                className='submit-button'
                disabled={!isDirty || isSubmitInProgress}
                data-testid='submit-button'
              >
                {format(`${isSubmitInProgress ? 'app.submitting' : 'app.submit'}`)}
                {isSubmitInProgress && <FaSpinner className='spinner ml-3 inline' data-testid='submit-spinner' />}
              </button>
              <button
                type='button'
                onClick={() => {
                  onCancel()
                  setIsInEditMode(false)
                }}
                className='cancel-button'
                disabled={isSubmitInProgress}
                data-testid='cancel-button'
              >
                {format('app.cancel')}
              </button>
            </div>
          </div>
        ) : displayModeBody
      }
    </div>
  )
}

export default EditableSection
