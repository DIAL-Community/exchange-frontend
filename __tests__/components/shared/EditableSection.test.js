import { fireEvent, screen } from '@testing-library/react'
import EditableSection from '../../../components/shared/EditableSection'
import { mockRouterImplementation, render } from '../../test-utils'

// Mock next-router calls.
jest.mock('next/dist/client/router')

describe('Unit test for the Editable Section component.', () => {
  const EDIT_BUTTON_TEST_ID = 'edit-button'
  const SUBMIT_BUTTON_TEST_ID = 'submit-button'
  const CANCEL_BUTTON_TEST_ID = 'cancel-button'
  const SUBMIT_SPINNER_TEST_ID = 'submit-spinner'
  const mockSectionHeader = 'Mock Section Header'
  const mockEditModeBody = 'edit mode body'
  const mockDisplayModeBody = 'display mode body'
  const mockOnSubmit = jest.fn()
  const mockOnCancel = jest.fn()

  beforeAll(mockRouterImplementation)

  test('Should match snapshot - without edit permission.', () => {
    const { container } = render(
      <EditableSection
        canEdit={false}
        sectionHeader={mockSectionHeader}
        editModeBody={mockEditModeBody}
        displayModeBody={mockDisplayModeBody}
      />
    )
    expect(container).toMatchSnapshot()
  })

  test('Should match snapshot - with edit permission.', () => {
    const { container } = render(
      <EditableSection
        canEdit={true}
        sectionHeader={mockSectionHeader}
        editModeBody={mockEditModeBody}
        displayModeBody={mockDisplayModeBody}
      />
    )
    expect(container).toMatchSnapshot()
  })

  test('Should switch to edit mode and back to display mode on "Cancel" button click.', () => {
    const { container, getByTestId } = render(
      <EditableSection
        canEdit={true}
        sectionHeader={mockSectionHeader}
        editModeBody={mockEditModeBody}
        displayModeBody={mockDisplayModeBody}
        onCancel={mockOnCancel}
      />
    )
    expect(screen.getByText(mockDisplayModeBody)).toBeInTheDocument()
    expect(screen.queryByText(mockEditModeBody)).toBeNull()

    fireEvent.click(getByTestId(EDIT_BUTTON_TEST_ID))
    expect(screen.queryByText(mockDisplayModeBody)).toBeNull()
    expect(screen.getByText(mockEditModeBody)).toBeInTheDocument()
    expect(container).toMatchSnapshot()

    fireEvent.click(getByTestId(CANCEL_BUTTON_TEST_ID))
    expect(mockOnCancel).toBeCalled()
    expect(screen.getByText(mockDisplayModeBody)).toBeInTheDocument()
    expect(screen.queryByText(mockEditModeBody)).toBeNull()
  })

  test('Should "Submit" button be disabled if no changes are made.', () => {
    const { getByTestId } = render(
      <EditableSection
        canEdit={true}
        sectionHeader={mockSectionHeader}
        editModeBody={mockEditModeBody}
        displayModeBody={mockDisplayModeBody}
        isDirty={false}
        onSubmit={mockOnSubmit}
      />
    )
    fireEvent.click(getByTestId(EDIT_BUTTON_TEST_ID))
    const submitButton = getByTestId(SUBMIT_BUTTON_TEST_ID)
    expect(submitButton).toBeDisabled()
    fireEvent.click(submitButton)
    expect(mockOnSubmit).not.toBeCalled()
  })

  test('Should "Submit" button be enabled when changes are made.', () => {
    const { getByTestId } = render(
      <EditableSection
        canEdit={true}
        sectionHeader={mockSectionHeader}
        editModeBody={mockEditModeBody}
        displayModeBody={mockDisplayModeBody}
        isDirty={true}
        onSubmit={mockOnSubmit}
      />
    )
    fireEvent.click(getByTestId(EDIT_BUTTON_TEST_ID))
    const submitButton = getByTestId(SUBMIT_BUTTON_TEST_ID)
    expect(submitButton).not.toBeDisabled()
    fireEvent.click(submitButton)
    expect(mockOnSubmit).toBeCalled()
  })

  test('Should "Submit" button become disabled and spinner should appear on "Submit" button click.', () => {
    const { getByTestId } = render(
      <EditableSection
        canEdit={true}
        sectionHeader={mockSectionHeader}
        editModeBody={mockEditModeBody}
        displayModeBody={mockDisplayModeBody}
        isDirty={true}
        isMutating={true}
        onSubmit={mockOnSubmit}
      />
    )
    fireEvent.click(getByTestId(EDIT_BUTTON_TEST_ID))
    expect(getByTestId(SUBMIT_BUTTON_TEST_ID)).toHaveTextContent('Submit')
    expect(screen.queryByTestId(SUBMIT_SPINNER_TEST_ID)).toBeNull()

    fireEvent.click(getByTestId(SUBMIT_BUTTON_TEST_ID))
    expect(mockOnSubmit).toBeCalled()
    expect(getByTestId(SUBMIT_BUTTON_TEST_ID)).toHaveTextContent('Submitting')
    expect(screen.queryByTestId(SUBMIT_SPINNER_TEST_ID)).toBeInTheDocument()
  })
})
