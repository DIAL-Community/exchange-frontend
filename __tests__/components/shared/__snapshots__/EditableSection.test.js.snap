// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for the Editable Section component. Should match snapshot - with edit permission. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="mt-12"
    data-testid="editable-section"
  >
    <div
      class="flex justify-between items-center mb-3"
    >
      <div
        class="card-title text-dial-gray-dark"
      >
        Mock Section Header
      </div>
      <button
        class="mr-2 inline-flex items-center gap-x-1.5 bg-dial-blue px-2 py-1 rounded-md text-white"
        data-testid="edit-button"
        type="button"
      >
        <img
          alt="Edit"
          class=""
          height="12px"
          src="/icons/edit.svg"
          width="12px"
        />
        <span
          class="text-sm"
        >
          Edit
        </span>
      </button>
    </div>
    display mode body
  </div>
</div>
`;

exports[`Unit test for the Editable Section component. Should match snapshot - without edit permission. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="mt-12"
    data-testid="editable-section"
  >
    <div
      class="flex justify-between items-center mb-3"
    >
      <div
        class="card-title text-dial-gray-dark"
      >
        Mock Section Header
      </div>
    </div>
    display mode body
  </div>
</div>
`;

exports[`Unit test for the Editable Section component. Should switch to edit mode and back to display mode on "Cancel" button click. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="mt-12"
    data-testid="editable-section"
  >
    <div
      class="flex justify-between items-center mb-3"
    >
      <div
        class="card-title text-dial-gray-dark"
      >
        Mock Section Header
      </div>
    </div>
    <div
      class="bg-edit p-6"
    >
      edit mode body
      <div
        class="flex justify-end mt-8 gap-3 text-xl"
      >
        <button
          class="submit-button"
          data-testid="submit-button"
          disabled=""
          type="submit"
        >
          Submit
        </button>
        <button
          class="cancel-button"
          data-testid="cancel-button"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
